Exponential Histogram support to the Prometheus exporter

fstab opened the pull request:
This PR adds support for exponential histograms in the Prometheus exporter as defined in the [Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md):

> An OpenTelemetry Exponential Histogram with a cumulative aggregation temporality MUST be converted to a Prometheus Native Histogram.

**About the Implementation**

The upstream [Prometheus Java Client library](https://github.com/prometheus/client_java) recently introduced a `prometheus-metrics-model`, which implements immutable read-only `MetricSnapshots`.

With this PR, we introduced an `Otel2PrometheusConverter` that converts OpenTelemetry's `MetricData` to Prometheus' `MetricSnapshots`.

That way, we can use the upstream HTTP exporter for exposing the `MetricSnapshots` in various exposition formats.

**Benefits**

* In Prometheus, exponential histograms require the Prometheus protobuf exposition format. By using the upstream exposition formats from the Prometheus library this comes for free, no need to re-implement the exposition format in `opentelemetry-java`.
* The Prometheus community are currently considering to work on a new version of OpenMetrics protobuf. By using the upstream exposition formats it will be trivial to add support for this in `opentelemetry-java` once this is done, because it will be sufficient to update the HTTP server dependency.
* The new `PrometheusMetricReader` class implements the Prometheus `MultiCollector` interface, which means it can be registered as a collector with the `PrometheusRegistry`. This makes it very easy to mix and match Prometheus instrumentation with OpenTelemetry SDK instrumentation when transitioning.

**Dependencies**

This PR adds a dependency to `io.prometheus:prometheus-metrics-httpserver`. However, the upstream module does not introduce any transitive dependencies outside the `io.prometheus` group. The Prometheus protobuf support is shaded, and does not introduce any external dependency. Here's the dependency tree as produced by `./gradlew :exporters:prometheus:dependencies`:

```
\--- io.prometheus:prometheus-metrics-exporter-httpserver:1.1.0
     \--- io.prometheus:prometheus-metrics-exporter-common:1.1.0
          +--- io.prometheus:prometheus-metrics-model:1.1.0
          \--- io.prometheus:prometheus-metrics-exposition-formats:1.1.0
               +--- io.prometheus:prometheus-metrics-model:1.1.0
               +--- io.prometheus:prometheus-metrics-config:1.1.0
               \--- io.prometheus:prometheus-metrics-shaded-protobuf:1.1.0
```

**Related**

This PR implements #5940 and fixes #5997. It will also make it easy to implement #6013.

asafm reviewed (changes requested):
As mentioned in the comment, the memory cost here is severe, and collides head to head with REUSABLE_DATA MemoryMode. I'm in the middle of implementing it, and when you make this change of data conversion, which cost lots of allocations, and worst: handing over control to a library outside the scope of this project, makes completing MemoryMode impossible.

My [suggestion](https://github.com/open-telemetry/opentelemetry-java/issues/5940#issuecomment-1793715629) as proposed in your proposal is to implement this with existing implementation. It's a small amount of code, 0 disruption to OTel Java SDK and most importantly: no external dependencies. My 2 comments in the proposal explained it in more detail.

fstab commented:
Thanks for looking into this @asafm.

> Implement this with existing implementation. It's a small amount of code, 0 disruption to OTel Java SDK

I don't think this is true. Adding support for Prometheus native histograms is not possible within the existing implementation, as they require a different data model and exposition format. This is a major refactoring in any case. Moreover, converting OpenTelemetry's exponential histograms to Prometheus' native histograms is not trivial.

> no external dependencies

I actually think the upstream dependency for the Prometheus endpoint is a good thing. 

First, it's not a random 3rd party dependency. Prometheus and OpenTelemetry are both CNCF projects, and we pledged mutual compatibility in the CNCF governance.
 
The previous implementation of the Prometheus exporter copied the upstream Prometheus code into `opentelemetry-java`. This will become increasingly difficult to maintain as we introduce complex metric types like Prometheus native histograms and more exposition formats like Prometheus protobuf and OpenMetrics protobuf.
 
Moreover, from a community perspective, using a common Prometheus metrics implementation will help keep the communities aligned. If we use a common Prometheus metrics implementation we will quickly recognize incompatibilities with future changes. If OpenTelemetry keeps using their own copy of the upstream code  there is a higher risk of slight differences getting unnoticed.

> cost lots of allocations

I'm not convinced that caching objects in memory results in better gc performance than allocating short-lived objects. After all, the gc marks live objects, and the more live objects you have the longer the mark phase.

However, if there is a way to improve gc performance in the Prometheus exporter it would be better to fix this upstream so that both communities can benefit rather than copying the code into `opentelemetry-java` and only fixing it there.

Anyway, I would appreciate if you could consider this PR. The only alternative I see is to duplicate the upstream implementation, copy it into `opentelemetry-java`, and have duplicate maintenance effort for all future developments like OpenMetrics protobuf. I don't think this is the better solution.

jack-berg commented:
> I actually think the upstream dependency for the Prometheus endpoint is a good thing.

> Having a dependency on https://github.com/prometheus/client_java is potentially a good thing.

I hadn't thought about this problem from that perspective. We actually take this approach with all the other built in exporters:
- ZipkinSpanExporter depends on `io.zipkin.reporter2:zipkin-reporter`, and `io.zipkin.reporter2:zipkin-sender-okhttp3`, and converts to the zipkin data model in [OtelToZipkinSpanTransformer](https://github.com/open-telemetry/opentelemetry-java/blob/c57ca65f7f1455e2f9d72f3d2d626aaf2f2f9a81/exporters/zipkin/src/main/java/io/opentelemetry/exporter/zipkin/OtelToZipkinSpanTransformer.java).
- JaegerGrpcSpanExpoter depends on the jaeger protos and converts to the jaeger data model in [SpanMarshaler](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/jaeger/src/main/java/io/opentelemetry/exporter/jaeger/SpanMarshaler.java). 
- JaegerThriftSpanExporter depends on `io.jaegertracing:jaeger-client` and converts to the jaeger data model in [Adapter](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/jaeger-thrift/src/main/java/io/opentelemetry/exporter/jaeger/thrift/Adapter.java).

However, prometheus is the exception. As @fstab points out, the "The previous implementation of the Prometheus exporter copied the upstream Prometheus code into opentelemetry-java", the current OpenTelemetry code was based on the prometheus [TextFormat](https://github.com/prometheus/client_java/blob/parent-0.16.0/simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java) as referenced in [Serializer](https://github.com/open-telemetry/opentelemetry-java/blob/c57ca65f7f1455e2f9d72f3d2d626aaf2f2f9a81/exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java#L68). But if you look at #4178 where the prometheus serialization was originally added, part of the intent was definitely performance oriented, to avoid extra allocations:

> It's nice to also get some performance by directly serializing given how huge they can be (in particular, lots of arrays allocated for histograms in the adapter code that are eliminated). Many Java apps have a strong correlation between full GC and Prometheus scrapes :)

I think the philosophical question we need to answer is: Do we have an obligation to hold ourselves to different or higher performance bar for the built in exporters we provider than the upstream implementations? I think probably no.

But I still want to have a good answer for performance centric use cases. In my opinion, the answer for this is to lean into optimizing OTLP (as I discussed in this [comment](https://github.com/open-telemetry/opentelemetry-java/issues/5940#issuecomment-1783121135)), and depending on prometheus support for native OTLP ingest. @asafm [counters](https://github.com/open-telemetry/opentelemetry-java/issues/5940#issuecomment-1793715629) that the road for native OTLP ingest could be a long one, discussing other systems part of the prometheus ecosystem like M3DB, Cortex, and VictoriaMetrics. I did a quick search, and it looks like OTLP support [is not on the radar of M3DB](https://github.com/m3db/m3/issues?q=is%3Aissue+is%3Aclosed+otlp), is [likely to be supported in Cortex](https://github.com/cortexproject/cortex/issues/4981), and ❗[is supported in VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2424)❗, so I think it might not actually take that long.

fstab commented:
Thanks @jack-berg.

> OTLP support in the Prometheus ecosystem

For completeness:
* [Mimir](https://github.com/grafana/mimir/) supports native OTLP over HTTP.
* The Prometheus team are working on [making Prometheus as an OTel native metrics backend](https://docs.google.com/document/d/1NGdKqcmDExynRXgC_u1CDtotz9IUdMrq2yyIq95hl70/edit#heading=h.2ryqf1yl1v4o), which includes adding an OTLP endpoint to the Prometheus server.

asafm commented:
Thanks @jack-berg for the detailed response. It did add the context I lacked here.
After carefully reading your response, and thinking about it taking in all the context provided, I agree with your approach, which basically says: Let's refactor OTLP exporters to be almost zero memory allocations, since OTel owns and controls this protocol. If you are performance sensitive and decided to use REUSABLE_DATA memory mode, you should opt-in for OTLP exporters. If you do work with Prometheus database specifically, you can ship it to OTel collector and have it serve it Prometheus format. If you work with others like M3DB or Cortex, the collector can ship it to them using remote write protocol.

In light of having the other protocols use a 3rd party library, it makes sense for Prometheus to follow suit.

My personal preference, in an answer to your philosophical question is a bit different, but it doesn't change the fact that I agree with the approach chosen. OTel, as it stands today, IMO is x10 a better metrics library than any other, at least in the Java ecosystem. As such, it can be even better by having its implementation set the higher bar performance wise. It's a bit like the Apple ecosystem - you have everything on the chain, hence it's ultimately performs better, with the disadvantages it brings.  

One point that also made me lean in to your approach is that eventually the industry would consolidate into a single protocol and libraries. OTLP protocol and OTel SDK will become the defacto standard, hence the usage of Prometheus exporter would eventually decline and will not be the go-to exporter.

Practically speaking, it means in Apache Pulsar documentation, I'll recommend not configuring Prometheus exporter and opting for OTLP exporters, whether it's targeted at  OTel collector or directly to a vendor supporting OTLP.

jack-berg commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusMetricReaderTest.java:
```diff
@@ -0,0 +1,1037 @@
+    StringBuilder regex = new StringBuilder();
+    for (int i = 0; i < parts.length; i++) {
+      regex.append(Pattern.quote(parts[i]));
+      if (i <= parts.length - 2) {
```
Why only if `i <= parts.length - 2`?

fstab replied:
Changed to a `stream()` based implementation which is easier to read.

jack-berg commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusMetricReaderTest.java:
```diff
@@ -0,0 +1,1037 @@
+  private Tracer tracer;
+
+  @BeforeEach
+  public void setUp() {
```
#nit: Couple of requests to align with style of the repo:
- Make the class package private
- Make setup package private
- Make all the test methods package private
- Omit the `test` prefix in all the test method names

jack-berg commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java:
```diff
@@ -21,6 +20,7 @@ public final class PrometheusHttpServerBuilder {
 
   private String host = DEFAULT_HOST;
   private int port = DEFAULT_PORT;
+  private PrometheusRegistry prometheusRegistry = PrometheusRegistry.defaultRegistry;
```
This means that by default, PrometheusHttpServer will serve all metrics collected from `MetricCollector` instances registered with the default prometheus registry, rather than only metrics from SdkMeterProvider?

If so, I don't think that should be the default behavior. I think the default registry should be `new PrometheusRegistry`. 

I'm not sure its a good idea to allow users to set their own registry, since doing so could cause conflicts which `SdkMeterProvider` is not responsible for. @fstab what happens if multiple `MetricCollectors` are registered with the default registry and return conflicting MetricSnapshots?

fstab replied:
I changed the default to `new PrometheusRegistry()`.

However, I think it would be good to allow users to set their own registry, because that enables an easy transition path between the Prometheus library and the OTel SDK. You can simply use both libraries and register all metrics with the same Prometheus registry.

jack-berg commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServer.java:
```diff
@@ -166,112 +113,11 @@ public void close() {
-        exchange.getResponseBody().write(RESPONSE);
-      }
-      exchange.close();
-    }
```
Nice to see all this code go away 👍

jack-berg commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java:
```diff
@@ -46,21 +46,20 @@ public PrometheusHttpServerBuilder setExecutor(ExecutorService executor) {
   }
 
+  /** Sets the {@link PrometheusRegistry} to be used for {@link PrometheusHttpServer}. */
+  public PrometheusHttpServerBuilder setPrometheusRegistry(PrometheusRegistry prometheusRegistry) {
```
@jkwatson you may be interested in this. May remove the need for [prometheus client bridge](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/prometheus-client-bridge).

jack-berg commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelper.java:
```diff
@@ -5,208 +5,84 @@
- * @see <a href="https://prometheus.io/docs/practices/naming/#base-units">Prometheus best practices
- *     for units</a>
- */
-final class PrometheusUnitsHelper {
```
@psx95 would you care to review the changes to this file to see if any of the work in #5400 is undone?

psx95 replied:
Apologies for the delayed response @jack-berg , this notification got missed - I will take a look once and raise any concerns I find.

jack-berg replied:
Thanks @psx95!

jack-berg commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerTest.java:
```diff
@@ -260,28 +259,23 @@ void fetch_DuplicateMetrics() {
+            + "# TYPE foo_unit_total counter\n"
+            + "foo_unit_total{otel_scope_name=\"scope1\"} 1.0\n"
+            + "foo_unit_total{otel_scope_name=\"scope2\"} 2.0\n"
+            + "# TYPE target_info gauge\n"
```
This appears to fix #5693.

jack-berg commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusUnitsHelperTest.java:
```diff
@@ -1,108 +0,0 @@
-        // Null unit
-        Arguments.of(null, null),
-        // Misc - unit cleanup - no case match special char
-        Arguments.of("$1000", "1000"));
```
Can we adapt / retain this test to confirm the new behavior of `PrometheusUnitHelper`?

fstab replied:
Good point. I restored the test, but I also changed some methods where I think the test is wrong:

* `Arguments.of("m/m", "meters_per_minute")` -> `Arguments.of("m/min", "meters_per_minute")`
* `Arguments.of("W/w", "watts_per_week"),` -> `Arguments.of("W/wk", "watts_per_week")`
* `Arguments.of("TBy/y", "terabytes_per_year")` -> `Arguments.of("TBy/a", "terabytes_per_year")`
* `Arguments.of("g/g", "grams_per_g")` -> `Arguments.of("g/x", "grams_per_x")` (this is supposed to test an unknown "per" unit, but `g` is not unknown
* Removed all tests for removing special characters like `Arguments.of("__test $/°C", "test_per_C")` because I could not find any reference that these characters should be removed.

Please let me know if I got it wrong.

jack-berg replied:
This seems reasonable to me. @psx95 please take a look if you can.

psx95 replied:
I just wanted to confirm the following - 
 - In the [collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/aaaaa0c5555cab8388ab523e5ae107cc8f0d235a/pkg/translator/prometheus/normalize_name.go#L58C5-L58C15), there is a separate list of 'per' unit(s). Which is why the "per" test cases were using those units specifically.
  - I do not see "wk" as a specified unit in the collector. Are we okay with the discrepancy between collector and SDK ? 
  - In the collector, certain units - for instance "mo", "y" are only used for "per" unit cases i.e. only in the second part of the metric unit when split using "/".  IIUC, the current implementation might use the mappings for any part of the unit. Is this intentional ?
   - Regarding the removing unsupported symbols - it was from the [OTel Spec Doc, OTLP -> Prometheus](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1). It mentions that metric name is supposed to match the regex `[a-zA-Z_:]([a-zA-Z0-9_:])*` which makes symbols like `$` invalid and replaced with `_` instead.

jack-berg replied:
> I do not see "wk" as a specified unit in the collector. Are we okay with the discrepancy between collector and SDK ?

While I like the idea of symmetry with the collector, the `wk` unit is defined in [ucum](https://unitsofmeasure.org/ucum), so I'm ok including it. UCUM is defined in the [metric semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-units) for defining unit.

> In the [collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/aaaaa0c5555cab8388ab523e5ae107cc8f0d235a/pkg/translator/prometheus/normalize_name.go#L58C5-L58C15), there is a separate list of 'per' unit(s). Which is why the "per" test cases were using those units specifically.
> In the collector, certain units - for instance "mo", "y" are only used for "per" unit cases i.e. only in the second part of the metric unit when split using "/". IIUC, the current implementation might use the mappings for any part of the unit. Is this intentional ?

The specification doesn't include any language that restricts the conversion to the `*_per_*` version to a subset of units, so I think its fine. Makes it simpler to understand.

> Regarding the removing unsupported symbols - it was from the [OTel Spec Doc, OTLP -> Prometheus](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1). It mentions that metric name is supposed to match the regex [a-zA-Z_:]([a-zA-Z0-9_:])* which makes symbols like $ invalid and replaced with _ instead.

Metric names are still sanitized according to those rules. However, we now delegate to the prometheus client library: https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java#L448

> In places where the spec was ambiguous, the unit conversion implementation in https://github.com/open-telemetry/opentelemetry-java/pull/5400 was done to be consistent with the collector implementation.

If we've diverged in any ways that aren't discussed here, I'd love to resolve those inconsistencies.

psx95 replied:
I see, thanks for clearing this up. 

One thing I did notice was that the definition of illegal characters is different in [Prometheus Client Library](https://github.com/prometheus/client_java/blob/60f72b6d27d31a4a566670cebdfb2699073c705a/prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/PrometheusNaming.java#L16) and the regex defined in [OTel Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#otlp-metric-points-to-prometheus).  (*OTel spec does not allow `.`*)
 - Regex allowed by OTel Spec - `[a-zA-Z_:]([a-zA-Z0-9_:])*`
 - Regex allowed by Prometheus Client Library - `^[a-zA-Z_.:][a-zA-Z0-9_.:]+$`

psx95 replied:
> If we've diverged in any ways that aren't discussed here, I'd love to resolve those inconsistencies.

I think we should add back the tests that were testing for removal of special characters and leading/trailing `_` in the unit. 

Regarding special characters: While it is not explicitly specified in spec, it is implicitly indicated by the regex used in the prometheus client library.

Regarding leading/trailing `_`: Removing these are not mandated by the spec, but the implementation of unit conversion in the collector does not allow for any leading or trailing `_`.

fstab replied:
> illegal characters

Here's the context: The Prometheus community decided (see [dev summit 2023-09](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#heading=h.np6zum9h19xx)) to allow UTF-8 characters in metric and attribute names for better OTel compatibility.

This is not implemented in the Prometheus server yet, so today the Prometheus exposition format will only create metric names and attribute names matching `[a-zA-Z_:]([a-zA-Z0-9_:])*` (without the dot).

However, we extended the Prometheus client library to allow dots in metric names, hence the new regexp that allows dots.

The idea is: Today the dots will never be exposed, they are implicitly converted to underscores in the exposition format. However, users can use dots when creating metrics, so they can use metric names compliant with OpenTelemetry's semantic conventions today. As soon as the Prometheus server and exposition format supports dots in metric names they can enable dots without having to re-instrument their applications.

TL;DR: While dots are allowed when creating metrics, they are implicitly replaced with underscores when exposing metrics, at least until dots become officially supported by the Prometheus server and exposition format.

psx95 replied:
Thanks for the context @fstab.

jack-berg reviewed (commented):
Thanks a lot for working on this - I saw several things in the code that significantly simplify understanding the code and make it less brittle.

fstab commented on exporters/prometheus/build.gradle.kts:
```diff
@@ -12,12 +12,14 @@ dependencies {
   api(project(":sdk:metrics"))
 
   implementation(project(":sdk-extensions:autoconfigure-spi"))
+  implementation("io.prometheus:prometheus-metrics-exporter-httpserver:1.1.0")
```
done.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+  private static final String OTEL_SCOPE_VERSION = "otel_scope_version";
+  private static final long NANOS_PER_MILLISECOND = TimeUnit.MILLISECONDS.toNanos(1);
+
+  public Otel2PrometheusConverter() {
```
removed.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+   * @param otelScopeEnabled enable generation of the OpenTelemetry instrumentation scope info
+   *     metric and labels.
+   */
+  public Otel2PrometheusConverter(boolean otelScopeEnabled) {
```
Good point. I made it configurable via `PrometheusHttpServerBuilder` because the spec says:
> Prometheus exporters SHOULD provide a configuration option to disable the otel_scope_info metric and otel_scope_ labels.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+import javax.annotation.Nullable;
+
+/** Convert OpenTelemetry {@link MetricData} to Prometheus {@link MetricSnapshots}. */
+public class Otel2PrometheusConverter {
```
done

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+    this.otelScopeEnabled = otelScopeEnabled;
+  }
+
+  public MetricSnapshots convert(@Nullable Collection<MetricData> metricDataCollection) {
```
done

fstab commented:
Thanks a lot for the review @jack-berg. I worked through the comments, rebased, and pushed a commit with the code review fulfillment. Please review again.

jack-berg commented on dependencyManagement/build.gradle.kts:
```diff
@@ -60,6 +60,7 @@ val DEPENDENCIES = listOf(
   "com.google.guava:guava-beta-checker:1.0",
   "com.sun.net.httpserver:http:20070405",
   "com.tngtech.archunit:archunit-junit5:1.2.1",
+  "io.prometheus:prometheus-metrics-exporter-httpserver:1.1.0",
```
#nit: alpha order please!

fstab replied:
done

jack-berg commented on exporters/prometheus/build.gradle.kts:
```diff
@@ -12,14 +12,15 @@ dependencies {
   api(project(":sdk:metrics"))
 
   implementation(project(":sdk-extensions:autoconfigure-spi"))
+  implementation("io.prometheus:prometheus-metrics-exporter-httpserver")
```
What kind of compatibility guarantees does `prometheus-metrics-exporter-httpserver` offer? Couldn't find anything obvious on the [github repo](https://github.com/prometheus/client_java?tab=readme-ov-file) or [docs site](https://prometheus.github.io/client_java/).

fstab replied:
Good point, we should document this. I'll put it on my TODO list. All `1.x` versions should be compatible, if there is a breaking change we should release `2.x`. I also think that the API is small enough that we won't need a `2.x` in the foreseeable future.

jack-berg commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,543 @@
+      case DOUBLE_SUM:
+        SumData<DoublePointData> doubleSumData = metricData.getDoubleSumData();
+        if (doubleSumData.getAggregationTemporality() == AggregationTemporality.DELTA) {
+          return null;
```
Maybe add a comment on these delta cases that they should never happen because `PrometheusMetricReader#getAggregationTemporality` returns `CUMULATIVE`.

fstab replied:
done

jack-berg reviewed (commented):
There are still a number of unaddressed comments (maybe they were hidden in the github UI or something?), but the only one which is non-trivial is the [question of whether to use the default registry by default](https://github.com/open-telemetry/opentelemetry-java/pull/6015#discussion_r1424588459).

fstab commented:
> There are still a number of unaddressed comments (maybe they were hidden in the github UI or something?), but the only one which is non-trivial is the [question of whether to use the default registry by default](https://github.com/open-telemetry/opentelemetry-java/pull/6015#discussion_r1424588459).

Oh my, I didn't expand the "hidden conversations". Sorry for that, I'll work through them asap.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+    return new HistogramSnapshot(metadata, Arrays.asList(data));
+  }
+
+  @SuppressWarnings("MethodCanBeStatic")
```
Removed all `@SuppressWarnings("MethodCanBeStatic")` and made the methods `static`.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+      MetricMetadata metadata,
+      InstrumentationScopeInfo scope,
+      Collection<LongPointData> dataPoints) {
+    GaugeDataPointSnapshot[] data = new GaugeDataPointSnapshot[dataPoints.size()];
```
Good point, the arrays were an artifact from previous refactoring, but with the current code `List` is better. I replaced the arrays with `ArrayList`.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+    String[] names =
+        new String[attributes.size() + numberOfScopeAttributes + additionalAttributes.length / 2];
+    String[] values = new String[names.length];
+    int[] pos = new int[] {0}; // using an array because we want to increment in a forEach() lambda.
```
Not sure. Maybe a performance specialist could create an example where `AtomicInteger` is slower in multi threaded scenarios because it syncs with main memory? Anyway, in this example I don't think it makes a difference.

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+  private Labels convertAttributes(
+      @Nullable InstrumentationScopeInfo scope,
+      Attributes attributes,
+      String... additionalAttributes) {
```
done

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+      int scale = histogramData.getScale();
+      if (scale < -4) {
+        // Scale < -4 is not supported in Prometheus. Histograms with scale < -4 are dropped.
+        return null;
```
done

fstab commented on exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java:
```diff
@@ -0,0 +1,547 @@
+
+  @Nullable
+  private Exemplar convertLongExemplar(List<LongExemplarData> exemplars) {
+    if (exemplars.size() == 0) {
```
done

fstab commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusMetricReaderTest.java:
```diff
@@ -0,0 +1,1037 @@
+    String expected =
+        ""
+            + "# TYPE requests_size_bytes counter\n"
+            + "# UNIT requests_size_bytes bytes\n"
```
Yes, the Prometheus metrics library will handle Unit metadata correctly.

Little fun fact: Prometheus has two text formats:
* Prometheus text format, which does not have units.
* OpenMetrics, which does have units.

Which one is used depends on the `Accepts` header in the scrape request. By default, the Prometheus server scrapes OpenMetrics format and gets units. However, if you visit a Prometheus endpoint with a Web browser you'll get Prometheus text format which doesn't have units.

Anyway, all of this is implemented in the Prometheus library so the OTel SDK doesn't need to care.

jack-berg replied:
> Anyway, all of this is implemented in the Prometheus library so the OTel SDK doesn't need to care.

Brilliant!

fstab commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusMetricReaderTest.java:
```diff
@@ -0,0 +1,1037 @@
+      }
+    }
+    Assertions.assertTrue(
+        Pattern.matches(regex.toString(), actual), "Expected: " + expected + "\nActual: " + actual);
```
done

fstab commented:
I worked through the rest of the comment, sorry for missing the hidden conversations previously.

jack-berg reviewed (approved):
Looks great to me. Thanks!

jack-berg commented:
Heads up @open-telemetry/java-approvers - I'd like to get this merged for the January 1.34.0 release so please leave any comments / feedback you have with enough time allotted to respond.

jack-berg commented:
Will merge tomorrow if there are no additional comments.

psx95 commented on exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/PrometheusMetricNameMapperTest.java:
```diff
@@ -1,163 +0,0 @@
-            metricData, PrometheusType.forMetric(metricData)));
-  }
-
-  private static Stream<Arguments> provideRawMetricDataForTest() {
```
IIUC, these test cases are still valid. If this is indeed the case, is there a way we could retain these test cases ?

jack-berg replied:
Check out #6138 where I restore these test cases.

psx95 commented:
@jack-berg, @fstab I left a couple comments (mostly around tests). Let me know if they make sense to you.

Apologies for the late response.
